src/Alluxio.cc

```diff
@@ -0,0 +1,659 @@
+           "alluxio/client/file/options/DeleteOptions", "defaults", 
+           "()Lalluxio/client/file/options/DeleteOptions;");
+
+     m_env.callMethod(&deleteOptionsSetRecursive, (jobject) deleteOptionsDefaults.l, 
+           "setRecursive", "(Z)Lalluxio/client/file/options/DeleteOptions;", 
+           (jboolean) recursive);
+
+     try
+     {
+        m_env.callMethod(&ret, m_obj, "delete", 
+              "(Lalluxio/AlluxioURI;Lalluxio/client/file/options/DeleteOptions;)V",
+              uri->getJObj(), (jobject) deleteOptionsSetRecursive.l);
+     }
+     catch (NativeException &e) 
+     {
+        std::string exceptionString;
+        jni::JavaThrowable *jthrow = e.detail();
+        jthrowable throwable = jthrow->getException();
+        m_env.throwableToString(throwable, exceptionString);
+        std::cout << "Caught exception: " << exceptionString << std::endl;
+        goto exit;
+     }
+  }
+
+exit:
+
+  delete uri;
+  return;
+}
+
+void AlluxioFileSystem::renameFile(const char *origPath, const char *newPath)
+{
+    jvalue ret;
+
+    jAlluxioURI origURI = AlluxioURI::newURI(origPath);
+    jAlluxioURI newURI = AlluxioURI::newURI(newPath);
+
+    m_env.callMethod(&ret, m_obj, "rename", 
+            "(Lalluxio/AlluxioURI;Lalluxio/AlluxioURI;)V", 
+            origURI->getJObj(), newURI->getJObj());
+
+    delete origURI;
+    delete newURI;
+}
+
+void AlluxioFileSystem::appendToFile(const char *path, void *buff, int length)
+{
+    const int bufferSize = 1000000;
+    const char* appendSuffix = ".append";
+    int bytesRead = bufferSize;
+    // FIXME: Change to std::vector
+    char* inputBuffer = (char *) calloc(bufferSize, 1);
+    int appendPathLength = strlen(path) + strlen(appendSuffix) + 1;
+
+    // FIXME: Change to std::vector
+    char* appendPath = (char *) malloc(appendPathLength);
+
+    strncpy(appendPath, path, appendPathLength);
+    strcat(appendPath,".append");
+
+    if (inputBuffer == NULL || appendPath == NULL)
+    {
+        throw std::bad_alloc();
+    }
+
+    jFileInStream origFile = openFile(path);
+
+    // FIXME: Create file should preserve file options of original file
+    jFileOutStream newFile = createFile(appendPath);
+
+    // Copy the original file over to the new file
+    while (bytesRead == bufferSize)
+    {
+        bytesRead = origFile->read(inputBuffer, bufferSize);
+
+        newFile->write(inputBuffer, bytesRead);
+    }
+
+    // Perform the append
+    newFile->write(buff, length);
+    origFile->close();
+    newFile->close();
+
+    // Delete original file
+    deletePath(path);
+
+    // NOTE: This is the critical section of this method.  If a failure
+    // occurs here, we will be left with a .append file and no original file.
+    
+    // TODO: we should add some defensive code in the open file path to
+    // perform this rename automatically if we only find a .append file
+    
+    // Rename file
+    renameFile(appendPath, path);
```

Comment:
Does alluxio allow rename a file to an existing file (atomically)? If so, deleting the original file can be avoided.